File: Contracts/src/main/java/org/cubexmc/contract/storage/ContractStorage.java
package org.cubexmc.contract.storage;

import org.bukkit.Bukkit;
import org.bukkit.configuration.ConfigurationSection;
import org.bukkit.configuration.InvalidConfigurationException;
import org.bukkit.configuration.file.YamlConfiguration;
import org.cubexmc.contract.ContractPlugin;
import org.cubexmc.contract.model.Asset;
import org.cubexmc.contract.model.Contract;
import org.cubexmc.contract.model.ContractEvent;
import org.cubexmc.contract.model.ContractStatus;
import org.cubexmc.contract.model.ContractType;
import org.cubexmc.contract.model.Participant;
import org.cubexmc.contract.model.ParticipantRole;
import org.cubexmc.contract.model.PayoutCondition;
import org.cubexmc.contract.model.PayoutRecipient;
import org.cubexmc.contract.model.PayoutRule;
import org.cubexmc.contract.model.ResolutionRule;

import java.io.File;
import java.io.IOException;
import java.math.BigDecimal;
import java.nio.file.AtomicMoveNotSupportedException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.StandardCopyOption;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import java.util.Objects;
import java.util.Optional;
import java.util.UUID;
import java.util.logging.Logger;

public final class ContractStorage {
    private final File file;
    private final File backupFile;
    private final File dataFolder;
    private final Logger logger;
    private final Map<String, Contract> contracts = new LinkedHashMap<>();
    private volatile boolean dirty = false;

    public ContractStorage(ContractPlugin plugin) {
        this(new File(plugin.getDataFolder(), "contract.yml"), plugin.getLogger());
    }

    ContractStorage(File file, Logger logger) {
        this.file = file;
        this.backupFile = new File(file.getParentFile(), file.getName() + ".bak");
        this.dataFolder = file.getParentFile();
        this.logger = logger;
    }

    public void markDirty() {
        dirty = true;
    }

    public boolean isDirty() {
        return dirty;
    }

    public void flushIfDirty() throws IOException {
        if (!dirty) {
            return;
        }
        save();
    }

    public void load() {
        // Read into a temp map first so a failed/recovered load never leaves the live map half-populated.
        Map<String, Contract> loaded = readContracts();
        contracts.clear();
        contracts.putAll(loaded);
    }

    private Map<String, Contract> readContracts() {
        if (!file.exists()) {
            return new LinkedHashMap<>();
        }
        try {
            return parseContracts(loadStrict(file));
        } catch (Exception primary) {
            logger.severe("contract.yml is unreadable: " + primary.getMessage());
            if (backupFile.exists()) {
                try {
                    Map<String, Contract> recovered = parseContracts(loadStrict(backupFile));
                    logger.warning("Recovered " + recovered.size() + " contracts from " + backupFile.getName() + ".");
                    return recovered;
                } catch (Exception backup) {
                    logger.severe("Backup " + backupFile.getName() + " is also unreadable: " + backup.getMessage());
                }
            }
            // Refusing to start empty protects escrowed funds: an empty map would be flushed over the only
            // copy of the data and orphan every player's withdrawn money. Force the admin to restore a backup.
            throw new IllegalStateException(
                "contract.yml and its backup are both unreadable; refusing to start with an empty contract set.",
                primary);
        }
    }

    private YamlConfiguration loadStrict(File source) throws IOException, InvalidConfigurationException {
        YamlConfiguration yaml = new YamlConfiguration();
        yaml.load(source);
        return yaml;
    }

    private Map<String, Contract> parseContracts(YamlConfiguration yaml) {
        Map<String, Contract> result = new LinkedHashMap<>();
        ConfigurationSection root = yaml.getConfigurationSection("contracts");
        if (root == null) {
            return result;
        }
        for (String id : root.getKeys(false)) {
            ConfigurationSection section = root.getConfigurationSection(id);
            if (section == null) {
                continue;
            }
            try {
                Contract contract = readContract(id, section);
                result.put(contract.id(), contract);
            } catch (RuntimeException ex) {
                logger.warning("Skipping malformed contract " + id + ": " + ex.getMessage());
            }
        }
        return result;
    }

    public void save() throws IOException {
        if (dataFolder != null && !dataFolder.exists()) {
            dataFolder.mkdirs();
        }
        YamlConfiguration yaml = new YamlConfiguration();
        ConfigurationSection root = yaml.createSection("contracts");
        for (Contract contract : contracts.values()) {
            writeContract(root.createSection(contract.id()), contract);
        }
        saveAtomically(yaml);
        dirty = false;
    }

    /**
     * Persist the financial database crash-safely: serialize to a temp file, roll the previous good file
     * into {@code contract.yml.bak}, then atomically swap the temp file in. A crash mid-write can therefore
     * never truncate {@code contract.yml} and orphan escrowed funds.
     */
    private void saveAtomically(YamlConfiguration yaml) throws IOException {
        Path target = file.toPath();
        Path dir = target.getParent();
        if (dir != null) {
            Files.createDirectories(dir);
        }
        Path temp = Files.createTempFile(dir, "contract", ".tmp");
        try {
            yaml.save(temp.toFile());
            if (Files.exists(target)) {
                Files.copy(target, backupFile.toPath(), StandardCopyOption.REPLACE_EXISTING);
            }
            try {
                Files.move(temp, target, StandardCopyOption.REPLACE_EXISTING, StandardCopyOption.ATOMIC_MOVE);
            } catch (AtomicMoveNotSupportedException ex) {
                Files.move(temp, target, StandardCopyOption.REPLACE_EXISTING);
            }
        } finally {
            Files.deleteIfExists(temp);
        }
    }

    public void put(Contract contract) {
        contracts.put(contract.id(), contract);
    }

    public void remove(String id) {
        contracts.remove(id);
    }

    public Optional<Contract> findByPrefix(String input) {
        String normalized = input.replace("#", "").toLowerCase(Locale.ROOT);
        List<Contract> matches = contracts.values().stream()
            .filter(contract -> contract.id().toLowerCase(Locale.ROOT).startsWith(normalized)
                || contract.shortId().toLowerCase(Locale.ROOT).equals(normalized))
            .toList();
        return matches.size() == 1 ? Optional.of(matches.get(0)) : Optional.empty();
    }

    public Optional<Contract> findById(String id) {
        return Optional.ofNullable(contracts.get(id));
    }

    public List<Contract> all() {
        return contracts.values().stream()
            .sorted(Comparator.comparingLong(Contract::createdAt).reversed())
            .toList();
    }

    public List<Contract> openContracts() {
        return contracts.values().stream()
            .filter(contract -> contract.status() == ContractStatus.OPEN)
            .sorted(Comparator.comparingLong(Contract::createdAt).reversed())
            .toList();
    }

    private Contract readContract(String id, ConfigurationSection section) {
        if (section.contains("participants")) {
            return readNewFormat(id, section);
        }
        return readLegacyFormat(id, section);
    }

    private Contract readNewFormat(String id, ConfigurationSection section) {
        ContractType type = ContractType.valueOf(section.getString("type", "SERVICE"));
        ContractStatus status = ContractStatus.valueOf(section.getString("status", "OPEN"));
        ResolutionRule resolutionRule = ResolutionRule.valueOf(section.getString("resolution-rule", "OWNER_APPROVE"));

        List<Participant> participants = new ArrayList<>();
        Object pRaw = section.get("participants");
        if (pRaw instanceof List<?> list) {
            for (Object entry : list) {
                if (entry instanceof Map<?, ?> map) {
                    Participant p = Participant.fromMap(map);
                    if (p.uuid() != null) {
                        p.displayName(resolveName(p.uuid(), p.displayName()));
                    }
                    participants.add(p);
                }
            }
        }

        Participant arbiter = null;
        Object aRaw = section.get("arbiter");
        if (aRaw instanceof Map<?, ?> map) {
            arbiter = Participant.fromMap(map);
            if (arbiter.uuid() != null) {
                arbiter.displayName(resolveName(arbiter.uuid(), arbiter.displayName()));
            }
        }

        List<PayoutRule> payouts = new ArrayList<>();
        Object payoutsRaw = section.get("payouts");
        if (payoutsRaw instanceof List<?> list) {
            for (Object entry : list) {
                if (entry instanceof Map<?, ?> map) {
                    payouts.add(PayoutRule.fromMap(map));
                }
            }
        }

        Contract contract = new Contract(
            id,
            type,
            section.getString("title", ""),
            section.getString("description", ""),
            participants,
            arbiter,
            resolutionRule,
            payouts,
            status,
            section.getLong("created-at"),
            nullableLong(section, "accepted-at"),
            nullableLong(section, "submitted-at"),
            nullableLong(section, "completed-at"),
            section.getLong("expires-at"),
            section.getString("dispute-reason"),
            readEvents(section)
        );

        ConfigurationSection meta = section.getConfigurationSection("metadata");
        if (meta != null) {
            for (String key : meta.getKeys(false)) {
                contract.metadata.put(key, meta.getString(key, ""));
            }
        }
        return contract;
    }

    /** Migrate pre-abstraction yaml entries (owner-uuid / reward / commission-percent flat keys). */
    private Contract readLegacyFormat(String id, ConfigurationSection section) {
        UUID ownerUuid = UUID.fromString(Objects.requireNonNull(section.getString("owner-uuid")));
        String ownerName = resolveName(ownerUuid, section.getString("owner-name", "Unknown"));
        UUID contractorUuid = uuidOrNull(section.getString("contractor-uuid"));
        String contractorName = contractorUuid == null
            ? section.getString("contractor-name")
            : resolveName(contractorUuid, section.getString("contractor-name"));
        ContractStatus status = ContractStatus.valueOf(section.getString("status", "OPEN"));
        BigDecimal reward = readBigDecimal(section, "reward");
        BigDecimal creationFee = readBigDecimal(section, "creation-fee");
        BigDecimal commissionPercent = readBigDecimal(section, "commission-percent");

        Participant owner = new Participant(ParticipantRole.OWNER, ownerUuid, ownerName,
            List.of(Asset.money(reward)));
        Participant contractor = new Participant(ParticipantRole.CONTRACTOR, contractorUuid, contractorName,
            List.of());

        List<PayoutRule> rules = new ArrayList<>();
        BigDecimal payoutShare = new BigDecimal("100").subtract(commissionPercent);
        rules.add(new PayoutRule(PayoutCondition.SUCCESS, ParticipantRole.OWNER,
            PayoutRecipient.participant(ParticipantRole.CONTRACTOR), payoutShare));
        rules.add(new PayoutRule(PayoutCondition.SUCCESS, ParticipantRole.OWNER,
            PayoutRecipient.systemSink(), commissionPercent));
        rules.add(new PayoutRule(PayoutCondition.FAILURE, ParticipantRole.OWNER,
            PayoutRecipient.participant(ParticipantRole.OWNER), new BigDecimal("100")));
        rules.add(new PayoutRule(PayoutCondition.TIMEOUT, ParticipantRole.OWNER,
            PayoutRecipient.participant(ParticipantRole.OWNER), new BigDecimal("100")));

        Contract contract = new Contract(
            id,
            ContractType.SERVICE,
            section.getString("title", ""),
            section.getString("description", ""),
            List.of(owner, contractor),
            null,
            ResolutionRule.OWNER_APPROVE,
            rules,
            status,
            section.getLong("created-at"),
            nullableLong(section, "accepted-at"),
            nullableLong(section, "submitted-at"),
            nullableLong(section, "completed-at"),
            section.getLong("expires-at"),
            section.getString("dispute-reason"),
            readEvents(section)
        );
        contract.metadata.put("creation-fee", creationFee.toPlainString());
        contract.metadata.put("commission-percent", commissionPercent.toPlainString());
        return contract;
    }

    private List<ContractEvent> readEvents(ConfigurationSection section) {
        List<ContractEvent> result = new ArrayList<>();
        Object raw = section.get("events");
        if (!(raw instanceof List<?> list)) {
            return result;
        }
        for (Object item : list) {
            if (item instanceof Map<?, ?> map) {
                result.add(ContractEvent.fromMap(map));
            } else if (item instanceof String legacy) {
                result.add(ContractEvent.fromLegacyLine(legacy));
            }
        }
        return result;
    }

    private void writeContract(ConfigurationSection section, Contract contract) {
        section.set("type", contract.type().name());
        section.set("title", contract.title());
        section.set("description", contract.description());
        section.set("resolution-rule", contract.resolutionRule().name());

        List<Map<String, Object>> participants = new ArrayList<>();
        for (Participant p : contract.participants()) {
            participants.add(p.toMap());
        }
        section.set("participants", participants);

        if (contract.arbiter() != null) {
            section.set("arbiter", contract.arbiter().toMap());
        }

        List<Map<String, Object>> payouts = new ArrayList<>();
        for (PayoutRule rule : contract.payouts()) {
            payouts.add(rule.toMap());
        }
        section.set("payouts", payouts);

        section.set("status", contract.status().name());
        section.set("created-at", contract.createdAt());
        section.set("accepted-at", contract.acceptedAt());
        section.set("submitted-at", contract.submittedAt());
        section.set("completed-at", contract.completedAt());
        section.set("expires-at", contract.expiresAt());
        section.set("dispute-reason", contract.disputeReason());

        if (!contract.metadata.isEmpty()) {
            ConfigurationSection meta = section.createSection("metadata");
            for (Map.Entry<String, String> entry : contract.metadata.entrySet()) {
                meta.set(entry.getKey(), entry.getValue());
            }
        }

        List<Map<String, Object>> serialized = new ArrayList<>();
        for (ContractEvent event : contract.events()) {
            serialized.add(event.toMap());
        }
        section.set("events", serialized);
    }

    private BigDecimal readBigDecimal(ConfigurationSection section, String path) {
        Object raw = section.get(path);
        if (raw == null) {
            return BigDecimal.ZERO;
        }
        if (raw instanceof Number n) {
            return BigDecimal.valueOf(n.doubleValue());
        }
        try {
            return new BigDecimal(raw.toString());
        } catch (NumberFormatException ex) {
            return BigDecimal.ZERO;
        }
    }

    private String resolveName(UUID uuid, String fallback) {
        if (uuid == null) {
            return fallback;
        }
        try {
            String current = Bukkit.getOfflinePlayer(uuid).getName();
            return current != null ? current : fallback;
        } catch (Throwable t) {
            // Bukkit not initialized (unit tests) or a name lookup failure — keep the stored name.
            return fallback;
        }
    }

    private UUID uuidOrNull(String value) {
        if (value == null || value.isBlank()) {
            return null;
        }
        return UUID.fromString(value);
    }

    private Long nullableLong(ConfigurationSection section, String path) {
        return section.contains(path) ? section.getLong(path) : null;
    }
}

File: Contract/src/main/java/org/cubexmc/contract/storage/ContractStorage.kt
package org.cubexmc.contract.storage

import org.bukkit.Bukkit
import org.bukkit.configuration.ConfigurationSection
import org.bukkit.configuration.InvalidConfigurationException
import org.bukkit.configuration.file.YamlConfiguration
import org.bukkit.inventory.ItemStack
import org.cubexmc.contract.ContractPlugin
import org.cubexmc.contract.model.Asset
import org.cubexmc.contract.model.Contract
import org.cubexmc.contract.model.ContractEvent
import org.cubexmc.contract.model.ContractObjective
import org.cubexmc.contract.model.ContractStatus
import org.cubexmc.contract.model.ContractType
import org.cubexmc.contract.model.Participant
import org.cubexmc.contract.model.ParticipantRole
import org.cubexmc.contract.model.PayoutCondition
import org.cubexmc.contract.model.PayoutRecipient
import org.cubexmc.contract.model.PayoutRule
import org.cubexmc.contract.model.ResolutionRule
import java.io.File
import java.io.IOException
import java.math.BigDecimal
import java.nio.file.AtomicMoveNotSupportedException
import java.nio.file.Files
import java.nio.file.StandardCopyOption
import java.util.ArrayList
import java.util.Comparator
import java.util.LinkedHashMap
import java.util.Objects
import java.util.Optional
import java.util.UUID
import java.util.logging.Logger

class ContractStorage {
    private val file: File
    private val backupFile: File
    private val dataFolder: File?
    private val logger: Logger
    private val contracts: MutableMap<String, Contract> = LinkedHashMap()

    @Volatile
    private var dirty = false

    constructor(plugin: ContractPlugin) : this(File(plugin.dataFolder, "contract.yml"), plugin.logger)

    constructor(file: File, logger: Logger) {
        this.file = file
        this.backupFile = File(file.parentFile, file.name + ".bak")
        this.dataFolder = file.parentFile
        this.logger = logger
    }

    @Synchronized
    fun markDirty() {
        dirty = true
    }

    @Synchronized
    fun isDirty(): Boolean = dirty

    @Throws(IOException::class)
    @Synchronized
    fun flushIfDirty() {
        if (!dirty) {
            return
        }
        save()
    }

    @Synchronized
    fun load() {
        // Read into a temp map first so a failed/recovered load never leaves the live map half-populated.
        val loaded = readContracts()
        contracts.clear()
        contracts.putAll(loaded)
    }

    private fun readContracts(): Map<String, Contract> {
        if (!file.exists()) {
            return LinkedHashMap()
        }
        try {
            return parseContracts(loadStrict(file))
        } catch (primary: Exception) {
            logger.severe("contract.yml is unreadable: ${primary.message}")
            if (backupFile.exists()) {
                try {
                    val recovered = parseContracts(loadStrict(backupFile))
                    logger.warning("Recovered ${recovered.size} contracts from ${backupFile.name}.")
                    return recovered
                } catch (backup: Exception) {
                    logger.severe("Backup ${backupFile.name} is also unreadable: ${backup.message}")
                }
            }
            throw IllegalStateException(
                "contract.yml and its backup are both unreadable; refusing to start with an empty contract set.",
                primary,
            )
        }
    }

    @Throws(IOException::class, InvalidConfigurationException::class)
    private fun loadStrict(source: File): YamlConfiguration {
        val yaml = YamlConfiguration()
        yaml.load(source)
        return yaml
    }

    private fun parseContracts(yaml: YamlConfiguration): Map<String, Contract> {
        val result = LinkedHashMap<String, Contract>()
        val root = yaml.getConfigurationSection("contracts") ?: return result
        for (id in root.getKeys(false)) {
            val section = root.getConfigurationSection(id) ?: continue
            try {
                val contract = readContract(id, section)
                result[contract.id()] = contract
            } catch (ex: RuntimeException) {
                logger.warning("Skipping malformed contract $id: ${ex.message}")
            }
        }
        return result
    }

    @Throws(IOException::class)
    @Synchronized
    fun save() {
        if (dataFolder != null && !dataFolder.exists()) {
            dataFolder.mkdirs()
        }
        val yaml = YamlConfiguration()
        val root = yaml.createSection("contracts")
        for (contract in contracts.values) {
            writeContract(root.createSection(contract.id()), contract)
        }
        saveAtomically(yaml)
        dirty = false
    }

    /**
     * Persist the financial database crash-safely: serialize to a temp file, roll the previous good file
     * into {@code contract.yml.bak}, then atomically swap the temp file in. A crash mid-write can therefore
     * never truncate {@code contract.yml} and orphan escrowed funds.
     */
    @Throws(IOException::class)
    private fun saveAtomically(yaml: YamlConfiguration) {
        val target = file.toPath()
        val dir = target.parent
        if (dir != null) {
            Files.createDirectories(dir)
        }
        val temp = Files.createTempFile(dir, "contract", ".tmp")
        try {
            yaml.save(temp.toFile())
            if (Files.exists(target)) {
                Files.copy(target, backupFile.toPath(), StandardCopyOption.REPLACE_EXISTING)
            }
            try {
                Files.move(temp, target, StandardCopyOption.REPLACE_EXISTING, StandardCopyOption.ATOMIC_MOVE)
            } catch (ex: AtomicMoveNotSupportedException) {
                Files.move(temp, target, StandardCopyOption.REPLACE_EXISTING)
            }
        } finally {
            Files.deleteIfExists(temp)
        }
    }

    @Synchronized
    fun put(contract: Contract) {
        contracts[contract.id()] = contract
    }

    @Synchronized
    fun remove(id: String) {
        contracts.remove(id)
    }

    @Synchronized
    fun findByPrefix(input: String): Optional<Contract> {
        val normalized = input.replace("#", "").lowercase(java.util.Locale.ROOT)
        val matches = contracts.values.stream()
            .filter { contract ->
                contract.id().lowercase(java.util.Locale.ROOT).startsWith(normalized) ||
                    contract.shortId().lowercase(java.util.Locale.ROOT) == normalized
            }
            .toList()
        return if (matches.size == 1) Optional.of(matches[0]) else Optional.empty()
    }

    @Synchronized
    fun findById(id: String): Optional<Contract> = Optional.ofNullable(contracts[id])

    @Synchronized
    fun all(): List<Contract> =
        contracts.values.stream()
            .sorted(Comparator.comparingLong { contract: Contract -> contract.createdAt() }.reversed())
            .toList()

    @Synchronized
    fun openContracts(): List<Contract> =
        contracts.values.stream()
            .filter { contract -> contract.status() == ContractStatus.OPEN }
            .sorted(Comparator.comparingLong { contract: Contract -> contract.createdAt() }.reversed())
            .toList()

    private fun readContract(id: String, section: ConfigurationSection): Contract =
        if (section.contains("participants")) readNewFormat(id, section) else readLegacyFormat(id, section)

    private fun readNewFormat(id: String, section: ConfigurationSection): Contract {
        val type = ContractType.valueOf(section.getString("type", "SERVICE") ?: "SERVICE")
        val status = ContractStatus.valueOf(section.getString("status", "OPEN") ?: "OPEN")
        val resolutionRule = ResolutionRule.valueOf(section.getString("resolution-rule", "OWNER_APPROVE") ?: "OWNER_APPROVE")

        val participants = ArrayList<Participant>()
        val pRaw = section["participants"]
        if (pRaw is List<*>) {
            for (entry in pRaw) {
                if (entry is Map<*, *>) {
                    val p = Participant.fromMap(entry)
                    if (p.uuid() != null) {
                        p.displayName(resolveName(p.uuid(), p.displayName()))
                    }
                    participants.add(p)
                }
            }
        }

        var arbiter: Participant? = null
        val aRaw = section["arbiter"]
        if (aRaw is Map<*, *>) {
            arbiter = Participant.fromMap(aRaw)
            if (arbiter.uuid() != null) {
                arbiter.displayName(resolveName(arbiter.uuid(), arbiter.displayName()))
            }
        }

        val payouts = ArrayList<PayoutRule>()
        val payoutsRaw = section["payouts"]
        if (payoutsRaw is List<*>) {
            for (entry in payoutsRaw) {
                if (entry is Map<*, *>) {
                    payouts.add(PayoutRule.fromMap(entry))
                }
            }
        }

        val contract = Contract(
            id,
            type,
            section.getString("title", "") ?: "",
            section.getString("description", "") ?: "",
            participants,
            arbiter,
            resolutionRule,
            payouts,
            status,
            section.getLong("created-at"),
            nullableLong(section, "accepted-at"),
            nullableLong(section, "submitted-at"),
            nullableLong(section, "completed-at"),
            section.getLong("expires-at"),
            section.getString("dispute-reason"),
            readObjective(section),
            readDeliveryItems(section),
            readEvents(section),
            readRewardItems(section),
        )

        val meta = section.getConfigurationSection("metadata")
        if (meta != null) {
            for (key in meta.getKeys(false)) {
                contract.metadata[key] = meta.getString(key, "") ?: ""
            }
        }
        return contract
    }

    /** Migrate pre-abstraction yaml entries (owner-uuid / reward / commission-percent flat keys). */
    private fun readLegacyFormat(id: String, section: ConfigurationSection): Contract {
        val ownerUuid = UUID.fromString(Objects.requireNonNull(section.getString("owner-uuid")))
        val ownerName = resolveName(ownerUuid, section.getString("owner-name", "Unknown"))
        val contractorUuid = uuidOrNull(section.getString("contractor-uuid"))
        val contractorName = if (contractorUuid == null) {
            section.getString("contractor-name")
        } else {
            resolveName(contractorUuid, section.getString("contractor-name"))
        }
        val status = ContractStatus.valueOf(section.getString("status", "OPEN") ?: "OPEN")
        val reward = readBigDecimal(section, "reward")
        val creationFee = readBigDecimal(section, "creation-fee")
        val commissionPercent = readBigDecimal(section, "commission-percent")

        val owner = Participant(ParticipantRole.OWNER, ownerUuid, ownerName, listOf(Asset.money(reward)))
        val contractor = Participant(ParticipantRole.CONTRACTOR, contractorUuid, contractorName, emptyList())

        val rules = ArrayList<PayoutRule>()
        val payoutShare = BigDecimal("100").subtract(commissionPercent)
        rules.add(
            PayoutRule(
                PayoutCondition.SUCCESS,
                ParticipantRole.OWNER,
                PayoutRecipient.participant(ParticipantRole.CONTRACTOR),
                payoutShare,
            ),
        )
        rules.add(PayoutRule(PayoutCondition.SUCCESS, ParticipantRole.OWNER, PayoutRecipient.systemSink(), commissionPercent))
        rules.add(
            PayoutRule(
                PayoutCondition.FAILURE,
                ParticipantRole.OWNER,
                PayoutRecipient.participant(ParticipantRole.OWNER),
                BigDecimal("100"),
            ),
        )
        rules.add(
            PayoutRule(
                PayoutCondition.TIMEOUT,
                ParticipantRole.OWNER,
                PayoutRecipient.participant(ParticipantRole.OWNER),
                BigDecimal("100"),
            ),
        )

        val contract = Contract(
            id,
            ContractType.SERVICE,
            section.getString("title", "") ?: "",
            section.getString("description", "") ?: "",
            listOf(owner, contractor),
            null,
            ResolutionRule.OWNER_APPROVE,
            rules,
            status,
            section.getLong("created-at"),
            nullableLong(section, "accepted-at"),
            nullableLong(section, "submitted-at"),
            nullableLong(section, "completed-at"),
            section.getLong("expires-at"),
            section.getString("dispute-reason"),
            null,
            emptyList(),
            readEvents(section),
        )
        contract.metadata["creation-fee"] = creationFee.toPlainString()
        contract.metadata["commission-percent"] = commissionPercent.toPlainString()
        return contract
    }

    private fun readEvents(section: ConfigurationSection): List<ContractEvent> {
        val result = ArrayList<ContractEvent>()
        val raw = section["events"]
        if (raw !is List<*>) {
            return result
        }
        for (item in raw) {
            if (item is Map<*, *>) {
                result.add(ContractEvent.fromMap(item))
            } else if (item is String) {
                result.add(ContractEvent.fromLegacyLine(item))
            }
        }
        return result
    }

    private fun writeContract(section: ConfigurationSection, contract: Contract) {
        section["type"] = contract.type().name
        section["title"] = contract.title()
        section["description"] = contract.description()
        section["resolution-rule"] = contract.resolutionRule().name

        val participants = ArrayList<Map<String, Any?>>()
        for (p in contract.participants()) {
            participants.add(p.toMap())
        }
        section["participants"] = participants

        if (contract.arbiter() != null) {
            section["arbiter"] = contract.arbiter()?.toMap()
        }

        val payouts = ArrayList<Map<String, Any?>>()
        for (rule in contract.payouts()) {
            payouts.add(rule.toMap())
        }
        section["payouts"] = payouts

        section["status"] = contract.status().name
        section["created-at"] = contract.createdAt()
        section["accepted-at"] = contract.acceptedAt()
        section["submitted-at"] = contract.submittedAt()
        section["completed-at"] = contract.completedAt()
        section["expires-at"] = contract.expiresAt()
        section["dispute-reason"] = contract.disputeReason()
        val objective = contract.objective()
        if (objective != null) {
            section["objective"] = objective.toMap()
        }
        if (contract.hasDeliveryItems()) {
            section["delivery-items"] = contract.deliveryItems().map { it.serialize() }
        }
        if (contract.hasRewardItems()) {
            section["reward-items"] = contract.rewardItems().map { it.serialize() }
        }

        if (contract.metadata.isNotEmpty()) {
            val meta = section.createSection("metadata")
            for ((key, value) in contract.metadata) {
                meta[key] = value
            }
        }

        val serialized = ArrayList<Map<String, Any?>>()
        for (event in contract.events()) {
            serialized.add(event.toMap())
        }
        section["events"] = serialized
    }

    private fun readObjective(section: ConfigurationSection): ContractObjective? {
        val nested = section.getConfigurationSection("objective")
        if (nested != null) {
            val map = LinkedHashMap<String, Any?>()
            for (key in nested.getKeys(false)) {
                map[key] = nested[key]
            }
            return ContractObjective.fromMap(map)
        }
        val raw = section["objective"]
        return if (raw is Map<*, *>) ContractObjective.fromMap(raw) else null
    }

    private fun readDeliveryItems(section: ConfigurationSection): List<ItemStack> {
        return readItems(section, "delivery-items")
    }

    private fun readRewardItems(section: ConfigurationSection): List<ItemStack> {
        return readItems(section, "reward-items")
    }

    private fun readItems(section: ConfigurationSection, path: String): List<ItemStack> {
        val raw = section[path]
        if (raw !is List<*>) {
            return emptyList()
        }
        val result = ArrayList<ItemStack>()
        for (entry in raw) {
            try {
                when (entry) {
                    is ItemStack -> result.add(entry.clone())
                    is Map<*, *> -> {
                        val map = LinkedHashMap<String, Any>()
                        for ((key, value) in entry) {
                            if (key is String && value != null) {
                                map[key] = value
                            }
                        }
                        result.add(ItemStack.deserialize(map))
                    }
                }
            } catch (ex: RuntimeException) {
                logger.warning("Skipping malformed stored $path item: ${ex.message}")
            }
        }
        return result
    }

    private fun readBigDecimal(section: ConfigurationSection, path: String): BigDecimal {
        val raw = section[path] ?: return BigDecimal.ZERO
        if (raw is Number) {
            return BigDecimal.valueOf(raw.toDouble())
        }
        return try {
            BigDecimal(raw.toString())
        } catch (ex: NumberFormatException) {
            BigDecimal.ZERO
        }
    }

    private fun resolveName(uuid: UUID?, fallback: String?): String? {
        if (uuid == null) {
            return fallback
        }
        return try {
            val current = Bukkit.getOfflinePlayer(uuid).name
            current ?: fallback
        } catch (t: Throwable) {
            fallback
        }
    }

    private fun uuidOrNull(value: String?): UUID? {
        if (value.isNullOrBlank()) {
            return null
        }
        return UUID.fromString(value)
    }

    private fun nullableLong(section: ConfigurationSection, path: String): Long? =
        if (section.contains(path)) section.getLong(path) else null
}
